nautobot/core/nautobot_app/__init__.py

```diff
@@ -0,0 +1,169 @@
+class NavMenuTab:
+    permissions = []
+    groups = []
+
+    def __init__(self, name, permissions=None, groups=None, weight=1000):
+        self.name = name
+        self.weight = weight
+        if permissions is not None:
+            if type(permissions) not in (list, tuple):
+                raise TypeError("Permissions must be passed as a tuple or list.")
+            self.permissions = permissions
+        if groups is not None:
+            if type(groups) not in (list, tuple):
+                raise TypeError("Items must be passed as a tuple or list.")
+            self.groups = groups
+
+
+class NavMenuGroup:
+    permissions = []
+    items = []
+
+    def __init__(self, name, items=None, weight=1000):
+        self.name = name
+        self.weight = weight
+        self.groups = {}
+        if items is not None:
+            if type(items) not in (list, tuple):
+                raise TypeError("Items must be passed as a tuple or list.")
+            self.items = items
+
+
+class NavMenuItem:
+    """
+    This class represents a navigation menu item. This constitutes primary link and its text, but also allows for
+    specifying additional link buttons that appear to the right of the item in the van menu.
+
+    Links are specified as Django reverse URL strings.
+    Buttons are each specified as a list of NavMenuButton instances.
+    """
+
+    def __init__(self, link, link_text, permissions=None, buttons=None, weight=1000):
+        reverse(link)
+        self.link = link
+        self.link_text = link_text
+        self.weight = weight
+        if permissions is not None:
+            if type(permissions) not in (list, tuple):
+                raise TypeError("Permissions must be passed as a tuple or list.")
+            self.permissions = permissions
+        if buttons is not None:
+            if type(buttons) not in (list, tuple):
+                raise TypeError("Buttons must be passed as a tuple or list.")
+            self.buttons = buttons
+
+
+class NavMenuButton:
+    """
+    This class represents a button within a PluginMenuItem. Note that button colors should come from
+    ButtonColorChoices.
+    """
+
+    def __init__(
+        self,
+        link,
+        title,
+        icon_class,
+        button_class=CustomLinkButtonClassChoices.CLASS_DEFAULT,
+        permissions=None,
+        weight=1000,
+    ):
+        reverse(link)
+        self.link = link
+        self.title = title
+        self.icon_class = icon_class
+        self.weight = weight
+        if permissions is not None:
+            if type(permissions) not in (list, tuple):
+                raise TypeError("Permissions must be passed as a tuple or list.")
+            self.permissions = permissions
+        if button_class is not None:
+            if button_class not in CustomLinkButtonClassChoices.values():
+                raise ValueError("Button color must be a choice within ButtonColorChoices.")
+            self.button_class = button_class
```

Comment:
As I pointed out for `self.permissions`, any of the other instance variables you're setting, should be set outside of the `if` blocks to avoid hitting an `AttributeError` when they are accessed but were never set.